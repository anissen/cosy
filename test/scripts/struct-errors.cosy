
struct Point {
	mut x Num = 2
    let y = 4
    let a Array Num
}

struct Fake {
	mut x Num = 2
    mut a = '4'
}

{
    mut point = Point { x = 5, y = 13, a = [] } // ok
    print point
}

{
    mut point = Point { x = 5, y = 13, a = [], x = 3 } // fails (x initialized twice)
    print point
}

{
    mut point = Point { x = 5, y = 13 } // fails (a is not initialized)
    print point
}

{
    mut point = Point { a = [] } // ok
    point.x = 555
    point.y = 7 // fails (y is not mut)
    point.z = 8 // fails (z is not a member)
    point.w = fn() { // fails (w is not a member)
    	print 'hello'
    }
    print point
    print '{point.x}, {point.y}'
}

{
    mut p = Point { a = [] }
    p = Fake {} // fails (Point type != Fake type)
    print p
}

{
    struct Test {
        let a Array Num
        let p = Point { x = 4, y = 12 }
    }

    let test = Test { a = [], p = Point { x = 5, y = 13 } }
    print test
}
