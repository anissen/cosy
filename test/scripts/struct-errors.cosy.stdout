■ test/scripts/struct-errors.cosy, line 19:
17 | 
18 | {
19 |     mut point = Point { x = 5, y = 13, a = [], x = 3 } // fails (x initialized twice)
                                                    ^ Member already assigned in initializer.
20 |     print point
21 | }

■ test/scripts/struct-errors.cosy, line 24:
22 | 
23 | {
24 |     mut point = Point { x = 5, y = 13 } // fails (a is not initialized)
                     ^^^^^ Member "a" not initialized.
25 |     print point
26 | }

■ test/scripts/struct-errors.cosy, line 31:
29 |     mut point = Point { a = [] } // ok
30 |     point.x = 555
31 |     point.y = 7 // fails (y is not mut)
               ^ Member is not mutable.
32 |     point.z = 8 // fails (z is not a member)
33 |     point.w = fn() { // fails (w is not a member)

■ test/scripts/struct-errors.cosy, line 32:
30 |     point.x = 555
31 |     point.y = 7 // fails (y is not mut)
32 |     point.z = 8 // fails (z is not a member)
               ^ No member named "z" in struct of type Struct { a Array(Num), x Num, y Num }
33 |     point.w = fn() { // fails (w is not a member)
34 |     	print 'hello'

■ test/scripts/struct-errors.cosy, line 33:
31 |     point.y = 7 // fails (y is not mut)
32 |     point.z = 8 // fails (z is not a member)
33 |     point.w = fn() { // fails (w is not a member)
               ^ No member named "w" in struct of type Struct { a Array(Num), x Num, y Num }
34 |     	print 'hello'
35 |     }

■ test/scripts/struct-errors.cosy, line 42:
40 | {
41 |     mut p = Point { a = [] }
42 |     p = Fake {} // fails (Point type != Fake type)
         ^ Cannot assign Struct { a Str, x Num } to Struct { a Array(Num), x Num, y Num }
43 |     print p
44 | }

■ test/scripts/struct-errors.cosy, line 49:
47 |     struct Test {
48 |         let a Array Num
49 |         let p = Point { x = 4, y = 12 }
                     ^^^^^ Member "a" not initialized.
50 |     }
51 | 

■ test/scripts/struct-errors.cosy, line 52:
50 |     }
51 | 
52 |     let test = Test { a = [], p = Point { x = 5, y = 13 } }
                                       ^^^^^ Member "a" not initialized.
53 |     print test
54 | }

